django/forms/forms.py

```diff
@@ -23,12 +23,14 @@ class DeclarativeFieldsMetaclass(MediaDefiningClass):
     """Collect Fields declared on the base classes."""
     def __new__(mcs, name, bases, attrs):
         # Collect fields from current class.
-        current_fields = []
-        for key, value in list(attrs.items()):
-            if isinstance(value, Field):
-                current_fields.append((key, value))
-                attrs.pop(key)
-        attrs['declared_fields'] = dict(current_fields)
+        current_fields = {
+            key: value for key, value in attrs.items()
+            if isinstance(value, Field)
+        }
+        for key in current_fields:
+            attrs.pop(key)
+
+        attrs['declared_fields'] = current_fields
```

Comment:
Not sure of the motivation behind these changes but this could be reduced to the following?

```suggestion
        attrs['declared_fields'] = {
            key: attrs.pop(key) for key, value in attrs.items()
            if isinstance(value, Field)
        }
```

Reply:
Won't that complain because you'd be removing items while iterating?

Reply:
This raises exception `RuntimeError: dictionary changed size during iteration`.

Can be fixed as follows:
```python
attrs['declared_fields'] = {
    key: attrs.pop(key) for key, value in list(attrs.items())
    if isinstance(value, Field)
}
```

But this creates an intermediate list

Reply:
I don't get such an error on Python 3.8 at least, it's possible that an error might be raised on < 3.8 though?

Reply:
Never mind, looks it's the other way around. Works on 3.6, fails on 3.8.

Reply:
> Not sure of the motivation behind these changes but this could be reduced to the following?

The original code looks redundant. It's uses two intermediate lists: `current_fields` and `list(attrs.items())`.

Besides using dict comprehension more effective than `list.append`.

Reply:
> ~Never mind, looks it's the other way around. Works on 3.6, fails on 3.8.~ 

Strange, I have the same error on 3.6:
```bash
Python 3.6.8 (default, Feb  7 2019, 12:49:57) 
[GCC 6.3.0 20170516] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> attrs = {'1': 1, '2': '2', '3': 3}
>>> {key: attrs.pop(key) for key, value in attrs.items() if isinstance(value, int)}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 1, in <dictcomp>
RuntimeError: dictionary changed size during iteration
```

Reply:
[This variant](https://github.com/django/django/pull/13623#discussion_r515074580) looks more elegant?